### PR TITLE
Gh 180 pull to refresh

### PR DESCRIPTION
**Pull request by sche**
**What is done:**
- Pull to refresh for loadable views
- Show already loaded content if there is an error during pull to refresh
- Display snackbar with error reason
- Prepare HTTP client for handling internal error codes

**Known issues (to discuss)**
- It is possible to pull while the “Loading” screen is displayed. It overrides and destroys the Loading screen. 
- In this case for Tranasctions e.g. displayed “Transactions will appear here” screen
-  In case of Safe Setting it displays not full data
-  In case of Balances it displays empty screen

**To Clarify**
- Refresh start animation

### CODE REVIEW

**Comment by DmitryBespalov** [comment]
Unfortunately this breaks the "selection" of the transaction - after going back from the tx details I see this:

File: Multisig/Data/Services/Utils/HTTPClient.swift [hunk 1]
```diff
@@ -4,61 +4,95 @@
...
-    /// Creates new client with baseURL and logger
-    ///
-    /// - Parameters:
-    ///   - url: base url for creating all request urls
-    ///   - logger: logger for debugging and error purposes
-    public init(url: URL, logger: Logger? = nil) {
-        baseURL = url
-        self.logger = logger
+        struct ErrorDetails: Decodable {
+            let code: Int?
```
Comment (MouazAlzahabi) [inline Multisig/Data/Services/Utils/HTTPClient.swift]: Is there any need for code to be optional?

**Comment by DmitryBespalov** [comment]
another issues is that sometimes the pull to refresh does not show, because I'm scrolling the list for the transactions, so it doesn't show the control

File: Multisig/UI/UI Library/RefreshableScrollView.swift [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+        static func reduce(value: inout [PrefData], nextValue: () -> [PrefData]) {
+            value.append(contentsOf: nextValue())
+        }
+
+        typealias Value = [PrefData]
+    }
+}
+
+struct ActivityRep: UIViewRepresentable {
```
Comment (DmitryBespalov) [inline Multisig/UI/UI Library/RefreshableScrollView.swift]: we have the `ActivityIndicator` already

File: Multisig/UI/UI Library/RefreshableScrollView.swift [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+struct RefreshableKeyTypes {
+    enum ViewType: Int {
+        case movingView
+        case fixedView
+    }
+
+    struct PrefData: Equatable {
+        let vType: ViewType
```
Comment (DmitryBespalov) [inline Multisig/UI/UI Library/RefreshableScrollView.swift]: please use full names, it is cryptic

File: Multisig/UI/UI Library/RefreshableScrollView.swift [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+        let bounds: CGRect
+    }
+
+    struct PrefKey: PreferenceKey {
```
Comment (DmitryBespalov) [inline Multisig/UI/UI Library/RefreshableScrollView.swift:169]: same here (unclear name)

File: Multisig/UI/UI Library/RefreshableScrollView.swift [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+                    .frame(height: height).fixedSize()
+                    .offset(y: -height + (self.loading && self.frozen ? height : 0.0))
+                } else {
+                    Image(systemName: "arrow.down") // If not loading, show the arrow
+                        .resizable()
+                        .aspectRatio(contentMode: .fit)
+                        .frame(width: height * 0.25, height: height * 0.25).fixedSize()
+                        .padding(height * 0.375)
+                        .rotationEffect(rotation)
+                        .offset(y: -height + (loading && frozen ? +height : 0.0))
```
Comment (DmitryBespalov) [inline Multisig/UI/UI Library/RefreshableScrollView.swift]: is `+height` intended usage?

File: Multisig/UI/UI Library/RefreshableScrollView.swift [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+            // Update last scroll offset
+            self.previousScrollOffset = self.scrollOffset
+        }
+    }
+
+    func symbolRotation(_ scrollOffset: CGFloat) -> Angle {
+        // We will begin rotation, only after we have passed
+        // 60% of the way of reaching the threshold.
+        if scrollOffset < self.threshold * 0.60 {
```
Comment (DmitryBespalov) [inline Multisig/UI/UI Library/RefreshableScrollView.swift]: please extract the magic constant

File: Multisig/UI/UI Library/RefreshableScrollView.swift [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+            return .degrees(0)
+        } else {
+            // Calculate rotation, based on the amount of scroll offset
+            let h = Double(self.threshold)
+            let d = Double(scrollOffset)
+            let v = max(min(d - (h * 0.6), h * 0.4), 0)
+            return .degrees(180 * v / (h * 0.4))
```
Comment (DmitryBespalov) [inline Multisig/UI/UI Library/RefreshableScrollView.swift]: what is the formula here? the one-letter names are not understandable

File: Multisig/UI/UI Library/RefreshableScrollView.swift [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+    func symbolRotation(_ scrollOffset: CGFloat) -> Angle {
+        // We will begin rotation, only after we have passed
+        // 60% of the way of reaching the threshold.
+        if scrollOffset < self.threshold * 0.60 {
+            return .degrees(0)
+        } else {
+            // Calculate rotation, based on the amount of scroll offset
+            let h = Double(self.threshold)
+            let d = Double(scrollOffset)
+            let v = max(min(d - (h * 0.6), h * 0.4), 0)
```
Comment (DmitryBespalov) [inline Multisig/UI/UI Library/RefreshableScrollView.swift]: magic constants

File: Multisig/UI/Transactions/TransactionsList/TransactionsViewModel.swift [hunk 9]
```diff
@@ -57,10 +53,13 @@ class TransactionsViewModel: LoadableViewModel {
             .sink(receiveCompletion: { completion in
                 if case .failure(let error) = completion {
                     self.errorMessage = error.localizedDescription
+                    App.shared.viewState.show(message: error.localizedDescription)
```
Comment (MouazAlzahabi) [inline Multisig/UI/Transactions/TransactionsList/TransactionsViewModel.swift]: App.shared.viewState can be defined as private parameter

File: Multisig/UI/UI Library/RefreshableScrollView.swift [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+                self.refreshing = true
+            }
+
+            if self.refreshing {
+                // Crossing the threshold on the way up, we add a space at the top of the scrollview
+                if self.previousScrollOffset > self.threshold && self.scrollOffset <= self.threshold {
+                    self.frozen = true
+                }
+            } else {
+                // remove the sapce at the top of the scroll view
```
Comment (DmitryBespalov) [inline Multisig/UI/UI Library/RefreshableScrollView.swift]: typo "sapce"

File: Multisig/UI/Settings/Safe Settings/SafeSettingsViewModel.swift [hunk 11]
```diff
@@ -48,13 +39,18 @@ class SafeSettingsViewModel: LoadableViewModel {
...
                 self.isLoading = false
+                self.isRefreshing = false
             }, receiveValue: { response in
                 self.safe.update(from: response)
+                self.errorMessage = nil
             })
             .store(in: &subscribers)
     }
+
+    func refreshData() {}
```
Comment (DmitryBespalov) [inline Multisig/UI/Settings/Safe Settings/SafeSettingsViewModel.swift]: is this supposed to be empty?

**Review by MouazAlzahabi — APPROVED** [review]

File: Multisig/Data/Services/Utils/HTTPClient.swift [hunk 12]
```diff
@@ -4,61 +4,95 @@
...
+
+    enum Error: LocalizedError {
         case networkRequestFailed(URLRequest, URLResponse?, Data?)
         case entityNotFound(URLRequest, URLResponse?, Data?)
         case unprocessableEntity(URLRequest, URLResponse?, Data?)
         case unknownError(URLRequest, URLResponse?, Data?)
 
-        public var errorDescription: String? {
+        enum Message: String {
+            case networkRequestFailed = "No internet connection."
```
Comment (DmitryBespalov) [inline Multisig/Data/Services/Utils/HTTPClient.swift]: is this really the case that if a request fails, then it is always because of no internet?

File: Multisig/Data/Services/Utils/HTTPClient.swift [hunk 13]
```diff
@@ -4,61 +4,95 @@
...
+                    let details = try JSONDecoder().decode(ErrorDetails.self, from: data)
+                    guard let code = details.code else {
+                        #warning("TODO: Crashlytics")
+                        return Message.unexpectedError.rawValue
+                    }
+                    return Message(code: code).rawValue
+                } catch {
+                    #warning("TODO: Crashlytics")
+                    return Message.unexpectedError.rawValue
+                }
```
Comment (DmitryBespalov) [inline Multisig/Data/Services/Utils/HTTPClient.swift:81]: this code now looks complicated, not sure right now how to improve it

File: Multisig/Data/Services/Utils/HTTPClient.swift [hunk 14]
```diff
@@ -4,61 +4,95 @@
...
     var query: String? { return nil }
     var body: Data? { return nil }
     var headers: [String: String] { return [:] }
-
 }
 
 /// Synchronous http client
-public class HTTPClient {
-    /// Client error
+class HTTPClient {
```
Comment (DmitryBespalov) [inline Multisig/Data/Services/Utils/HTTPClient.swift:22]: maybe move the error handling out in another PR/ or another issue? This will make this PR more focused

**Review by DmitryBespalov — CHANGES_REQUESTED** [review]

**Comment by DmitryBespalov** [comment]
If this is merged with the "known issues", please create additional bug issues in the board

File: Multisig/Data/Services/Utils/HTTPClient.swift (hunk 1 above)
Comment (sche) [reply Multisig/Data/Services/Utils/HTTPClient.swift]: you are right, no need in optional

File: Multisig/UI/Transactions/TransactionsList/TransactionsViewModel.swift (hunk 9 above)
Comment (sche) [reply Multisig/UI/Transactions/TransactionsList/TransactionsViewModel.swift]: I think this way it is more explicit that we work with a global state

File: Multisig/UI/UI Library/RefreshableScrollView.swift (hunk 2 above)
Comment (sche) [reply Multisig/UI/UI Library/RefreshableScrollView.swift]: 👍

File: Multisig/UI/UI Library/RefreshableScrollView.swift (hunk 3 above)
Comment (sche) [reply Multisig/UI/UI Library/RefreshableScrollView.swift]: 👍

File: Multisig/UI/UI Library/RefreshableScrollView.swift (hunk 5 above)
Comment (sche) [reply Multisig/UI/UI Library/RefreshableScrollView.swift]: no :) copy-paste

File: Multisig/UI/UI Library/RefreshableScrollView.swift (hunk 6 above)
Comment (sche) [reply Multisig/UI/UI Library/RefreshableScrollView.swift]: ok

File: Multisig/UI/Settings/Safe Settings/SafeSettingsViewModel.swift (hunk 11 above)
Comment (sche) [reply Multisig/UI/Settings/Safe Settings/SafeSettingsViewModel.swift]: No, this is a leftover.

File: Multisig/UI/UI Library/RefreshableScrollView.swift (hunk 10 above)
Comment (sche) [reply Multisig/UI/UI Library/RefreshableScrollView.swift]: ok

File: Multisig/UI/UI Library/RefreshableScrollView.swift (hunk 4 above)
Comment (sche) [reply Multisig/UI/UI Library/RefreshableScrollView.swift:169]: What would be your proposal on the name here?

File: Multisig/Data/Services/Utils/HTTPClient.swift (hunk 12 above)
Comment (sche) [reply Multisig/Data/Services/Utils/HTTPClient.swift]: you are right. No internet connection description is provided by the system.
I will fix this message.

File: Multisig/Data/Services/Utils/HTTPClient.swift (hunk 14 above)
Comment (sche) [reply Multisig/Data/Services/Utils/HTTPClient.swift:22]: Yeah, I agree that there is a room for two PRs :( But I would not like to do it now.

File: Multisig/UI/UI Library/RefreshableScrollView.swift (hunk 7 above)
Comment (sche) [reply Multisig/UI/UI Library/RefreshableScrollView.swift]: improved

File: Multisig/UI/UI Library/RefreshableScrollView.swift (hunk 4 above)
Comment (DmitryBespalov) [reply Multisig/UI/UI Library/RefreshableScrollView.swift:169]: `Key`?

File: Multisig/UI/UI Library/RefreshableScrollView.swift [hunk 15]
```diff
@@ -16,7 +16,12 @@ struct RefreshableScrollView<Content: View>: View {
     @State private var frozen: Bool = false
     @State private var rotation: Angle = .degrees(0)
 
-    var threshold: CGFloat = 80
+    var threshold: CGFloat = 70
+
+    // We will begin rotation, only after we have passed
+    // 60% of the way of reaching the threshold.
+    var thresholdRubicon = 0.6
```
Comment (DmitryBespalov) [inline Multisig/UI/UI Library/RefreshableScrollView.swift:23]: Rubicon? Really? 😄 sorry, this is also not so explanatory, please make it boring 😄

File: Multisig/UI/UI Library/RefreshableScrollView.swift [hunk 16]
```diff
@@ -87,16 +87,20 @@ struct RefreshableScrollView<Content: View>: View {
...
             return .degrees(0)
         } else {
             // Calculate rotation, based on the amount of scroll offset
-            let h = Double(threshold)
-            let d = Double(scrollOffset)
-            let v = max(min(d - (h * thresholdRubicon), h * (1 - thresholdRubicon)), 0)
-            return .degrees(180 * v / (h * (1 - thresholdRubicon)))
+            let t = Double(threshold)
+            let tr = t * thresholdRubicon
+            let tri = t * (1 - thresholdRubicon)
```
Comment (DmitryBespalov) [inline Multisig/UI/UI Library/RefreshableScrollView.swift]: `t`, `tr`, `tri` didn't make it better either 😜 

How about:

```swift
// rename the `threshold` variable to `refreshTresholdInPoints` and `thresholdRubicon` in `relativeRotationThreshold`;

let rotationStartInPoints = Double(refreshTresholdInPoints) * relativeRotationThreshold
let rotationEndInPoints = Double(refreshTresholdInPoints) * (1 - relativeRotationThreshold)

if scrollOffset < rotationStartInPoints {
  return .degrees(0)
} else {
  let currentRotationOffset = max(min(Double(scrollOffset) - rotationStartInPoints, rotationEndInPoints), 0)
  let rotationAngle = Angle.degrees(180 * currentRotationOffset / rotationEndInPoints)
}
```

**Review by DmitryBespalov — CHANGES_REQUESTED** [review]

File: Multisig/UI/UI Library/RefreshableScrollView.swift (hunk 4 above)
Comment (sche) [reply Multisig/UI/UI Library/RefreshableScrollView.swift:169]: Then `PrefData` would be `Data` which is also not ideal.
I would leave it as it is for now if you don't mind.

File: Multisig/UI/UI Library/RefreshableScrollView.swift (hunk 16 above)
Comment (sche) [reply Multisig/UI/UI Library/RefreshableScrollView.swift]: ok. I changed.

File: Multisig/UI/App/Theme.swift [hunk 17]
```diff
@@ -39,4 +39,9 @@ class Theme: ObservableObject {
         UITableView.appearance().backgroundColor = tableViewBackgroundColor
     }
 
+    func resetRowsSelection() {
```
Comment (DmitryBespalov) [inline Multisig/UI/App/Theme.swift:42]: we need big ugly comment to justify why we need this

**Review by DmitryBespalov — APPROVED** [review]
